include/rmm/mr/device/binning_memory_resource.hpp

```diff
@@ -0,0 +1,220 @@
+  {
+    allocation_size = rmm::detail::align_up(allocation_size, allocation_alignment);
+
+    if (nullptr != bin_resource)
+      resource_bins_.insert({allocation_size, bin_resource});
+    else {
+      // If the bin already exists, do nothing.
+      if (resource_bins_.count(allocation_size) == 0) {
+        owned_bin_resources_.push_back(
+          std::make_unique<fixed_size_memory_resource<Upstream>>(upstream_mr_, allocation_size));
+        resource_bins_.insert({allocation_size, owned_bin_resources_.back().get()});
+      }
+    }
+  }
+
+ private:
+  /**
+   * @brief Get the memory resource for the requested size
+   *
+   * Chooses a memory_resource that allocates the smallest blocks at least as large as `bytes`.
+   *
+   * @param bytes Requested allocation size in bytes
+   * @return rmm::mr::device_memory_resource& memory_resource that can allocate the requested size.
+   */
+  device_memory_resource* get_resource(std::size_t bytes)
+  {
+    auto iter = resource_bins_.lower_bound(bytes);
+    return (iter != resource_bins_.cend()) ? iter->second
+                                           : static_cast<device_memory_resource*>(get_upstream());
+  }
+
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * @param bytes The size of the allocation
+   * @param stream Stream on which to perform allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+    return get_resource(bytes)->allocate(bytes, stream);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws nothing
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    auto res = get_resource(bytes);
+    if (res != nullptr) res->deallocate(p, bytes, stream);
+  }
+
+  /**
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws std::runtime_error if we could not get free / total memory
+   *
+   * @param stream the stream being executed on
+   * @return std::pair with available and free memory for resource
+   */
+  std::pair<std::size_t, std::size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    return std::make_pair(0, 0);
+  }
+
+  Upstream* upstream_mr_;  // The upstream memory_resource from which to allocate blocks.
+
+  std::vector<std::unique_ptr<fixed_size_memory_resource<Upstream>>> owned_bin_resources_;
+
+  std::map<std::size_t, device_memory_resource*> resource_bins_;
+};  // namespace mr
```

Comment:
```suggestion
};
```

Reply:
Thanks, fixed. I think this must be a bug in clang-format, I've seen it several times.